cloudigrade/account/views.py

```diff
@@ -25,7 +50,37 @@ def get_queryset(self):
             return self.queryset.filter(user=user)
         user_id = self.request.query_params.get('user_id', None)
         if user_id is not None:
-            return self.queryset.filter(user__id=int(user_id))
+            user_id = convert_param_to_int('user_id', user_id)
+            return self.queryset.filter(user__id=user_id)
+        return self.queryset
+
+
+class InstanceViewSet(viewsets.ReadOnlyModelViewSet):
+    """
+    List all or retrieve a single instance.
+
+    Do not allow to create, update, replace, or delete an instance at
+    this view because we currently **only** allow instances to be retrieved.
+    """
+
+    queryset = Instance.objects.all()
+    accounts = Account.objects.all()
+    serializer_class = serializers.InstancePolymorphicSerializer
+
+    def get_queryset(self):
+        """Get the queryset filtered to appropriate user."""
+        user = self.request.user
+        if not user.is_superuser:
+            accounts = self.accounts.filter(user=user)
+            account_ids = [account.id for account in accounts]
+            return self.queryset.filter(account__in=account_ids)
+        user_id = self.request.query_params.get('user_id', None)
+        if user_id is not None:
+            user_id = convert_param_to_int('user_id', user_id)
+            accounts = self.accounts.filter(user__id=int(user_id))
+            account_ids = [account.id for account in accounts]
+            return self.queryset.filter(account__in=account_ids)
```

Comment:
I'm pretty sure there's a cleaner way of doing this that doesn't require the "in list" like this, but I'll have to play with this and come back…

Reply:
I played around in the shell with some local data, and this seems to work as desired:

```
>>> models.Instance.objects.filter(account__user__id=30)
<PolymorphicQuerySet [<AwsInstance i-0bee3c1c7c65b518c>, <AwsInstance i-05f78714782d7979b>]>
>>> models.Instance.objects.filter(account__user__id=31)
<PolymorphicQuerySet []>
```

So, I'd suggest simplifying `get_queryset` to look something like this:

```python
def get_queryset(self):
    """Get the queryset filtered to appropriate user."""
    user = self.request.user
    if not user.is_superuser:
        return self.queryset.filter(account__user__id=user.id)
    user_id = self.request.query_params.get('user_id', None)
    if user_id is not None:
        user_id = convert_param_to_int('user_id', user_id)
        return self.queryset.filter(account__user__id=user_id) 
    return self.queryset 
```

Reply:
Great suggestion.  Done.